llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -123,14 +123,125 @@ repodefinition::DigestType toDigestType(pstore::index::digest D) {
   return Digest;
 }
 
+static void
+collectSectionXFixupName(pstore::database const &Db,
+                         pstore::repo::generic_section const &Section,
+                         StringSet<> &XFixupNames) {
+  for (auto Xfx : Section.xfixups()) {
+    LLVM_DEBUG(dbgs() << "    Adding XFixup name: "
+                      << pstore::indirect_string::read(Db, Xfx.name).to_string()
+                      << '\n');
+    XFixupNames.insert(
+        std::move(pstore::indirect_string::read(Db, Xfx.name).to_string()));
+
+  }
+}
+
+#define PSTORE_GENERIC_SECTION_KINDS                                           \
+  X(text)                                                                      \
+  X(data)                                                                      \
+  X(rel_ro)                                                                    \
+  X(mergeable_1_byte_c_string)                                                 \
+  X(mergeable_2_byte_c_string)                                                 \
+  X(mergeable_4_byte_c_string)                                                 \
+  X(mergeable_const_4)                                                         \
+  X(mergeable_const_8)                                                         \
+  X(mergeable_const_16)                                                        \
+  X(mergeable_const_32)                                                        \
+  X(read_only)                                                                 \
+  X(thread_data)
+
+static void collectFragmentXFixupNames(
+    const pstore::database &Repository,
+    std::shared_ptr<const pstore::repo::fragment> const &Fragment,
+    StringSet<> &XFixupNames) {
+  for (const pstore::repo::section_kind Kind : *Fragment) {
+    assert(Fragment->has_section(Kind));
+
+#define X(k)                                                                   \
+  case pstore::repo::section_kind::k: {                                        \
+    auto Section = Fragment->atp<pstore::repo::section_kind::k>();             \
+    collectSectionXFixupName(Repository, *Section, XFixupNames);               \
+    break;                                                                     \
+  }
+
+    switch (Kind) {
+      PSTORE_GENERIC_SECTION_KINDS
+    case pstore::repo::section_kind::debug_line:
+    case pstore::repo::section_kind::dependent:
+    case pstore::repo::section_kind::bss:
+    case pstore::repo::section_kind::thread_bss:
+    case pstore::repo::section_kind::debug_string:
+    case pstore::repo::section_kind::debug_ranges:
+    case pstore::repo::section_kind::last:
+    case pstore::repo::section_kind::interp:
+      break;
+    }
+#undef X
+  }
+}
+
+// Reurn the GO's status as a tuple<InRepository, InModeuleFragments, Key>:
+// InRepository: return true if the GO is in the existing repository.
+// InModeuleFragments: reurn true if the GO is in the new module fragments.
+// Key: return GO's digest.
```

Comment:
The function seems to do more than this comment suggests. It takes a series of contains by non-const reference (ModuleFragments, XFixupNames, PrunedDiscardableIfUnusedGos) and modifies at least some of them. Perhaps expanding this description would help me to understand what this is trying to do?

Reply:
I have added detailed description for this function.